Mathlib/Tactic/StacksAttribute.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+
+/-!
+# The `stacks` attribute
+
+This allows tagging of mathlib lemmas with the corresponding
+[Tags](https://stacks.math.columbia.edu/tags) from the Stacks Project.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Stacks
+
+/-- `Tag` is the structure that carries the data of a Stacks Projects tag and a corresponding
+Mathlib declaration. -/
+structure Tag where
+  /-- The Stacks Project tag. -/
+  tag : String
+  /-- The name of the declaration with the given tag. -/
+  declName : Name
+  deriving BEq, Hashable
+
+/-- Defines the `tagExt` extension for adding a `HashSet` of `Tag`s
+to the environment. -/
+initialize tagExt : SimplePersistentEnvExtension Tag (HashSet Tag) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addTagEntry tag declName` takes as input the `String` `tag` and the `Name` `declName` of
+a declaration with the `stacks` attribute.
+It extends the `Tag` environment extension with the data `tag, declName`.
+-/
+def addTagEntry {m : Type → Type} [MonadEnv m] (tag : String) (declName : Name) : m Unit :=
+  modifyEnv (tagExt.addEntry · { tag := tag, declName := declName })
+
+/--
+The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
+This allows `044Q3` and `GH3F6` as possibilities.
+-/
+declare_syntax_cat stackTag
+
+@[inherit_doc Parser.Category.stackTag]
+syntax (num)? (ident)? : stackTag
+
+/-- The `stacks` attribute.
+Use it as `@[stacks TAG "Optional comment"]`.
+The `TAG` is mandatory.
+
+See the [Tags page](https://stacks.math.columbia.edu/tags) in the Stacks project for more details.
+-/
+syntax (name := stacks) "stacks " (stackTag)? (ppSpace str)? : attr
+
+initialize Lean.registerBuiltinAttribute {
+  name := `stacks
+  descr := "Apply a Stacks project tag to a theorem."
+  add := fun decl stx _attrKind => Lean.withRef stx do
+    -- check that the tag consists of 4 characters and
+    -- that only digits and uppercase letter are present
+    let tag := stx[1]
+    match tag.getSubstring? with
+      | none => logWarning "Please, enter a Tag after `stacks`."
+      | some str =>
+        let str := str.toString.trimRight
+        if str.length != 4 then
+          logWarningAt tag
+            m!"Tag '{str}' is {str.length} characters long, but it should be 4 characters long"
+        else if 2 ≤ (str.split (fun c => (!c.isUpper) && !c.isDigit)).length then
+          logWarningAt tag m!"Tag '{str}' should only consist of digits and uppercase letters"
+        else match stx with
+          | `(attr| stacks $_:stackTag $_:str) => addTagEntry str decl
```

Comment:
Uh oh. Are we keeping track of the comments at all?

Reply:
I thought that the comments would be discarded: should the `Stacks.Tag` record them as well?

Reply:
Yeah, I think that was why Kim asked for them. To record discrepancies between the two statements.

Reply:
Comments added!